File: examples/example-github-java-connector/java-module/src/main/java/com/snowflake/connectors/sdk/example_github_connector/ingestion/Ingestion.java
package com.snowflake.connectors.sdk.example_github_connector.ingestion;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.snowflake.connectors.sdk.common.DefaultKeyValueTable;
import com.snowflake.connectors.sdk.example_github_connector.model.GithubIssue;
import com.snowflake.snowpark_java.Row;
import com.snowflake.snowpark_java.SaveMode;
import com.snowflake.snowpark_java.Session;
import com.snowflake.snowpark_java.types.DataTypes;
import com.snowflake.snowpark_java.types.StructField;
import com.snowflake.snowpark_java.types.StructType;
import com.snowflake.snowpark_java.types.Variant;

import java.net.http.HttpResponse;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.snowflake.connectors.sdk.example_github_connector.application.Infrastructure.APP_ROLE;

public class Ingestion {
    private static final ObjectMapper objectMapper = new ObjectMapper();

    public String ingest_resource(Session session, String resourceName) {
        var table = new DefaultKeyValueTable(session, "STATE.APP_CONFIGURATION");
        String destDatabaseName = table.fetch("dest_database_name").asString();
        String destTableName = destDatabaseName + "." + "DEST_SCHEMA" + "." + resourceName;

        String url = String.format("https://api.github.com/repos/apache/%s/issues", resourceName);
        String query = String.format("SELECT rs.value AS result FROM LATERAL FLATTEN(INPUT => parse_json(PUBLIC.FETCH_DATA('%s'))) rs", url);
        session.sql(query).write().mode(SaveMode.Append).saveAsTable(destTableName);

        session.sql(String.format("GRANT SELECT ON TABLE %s TO APPLICATION ROLE %s", destTableName, APP_ROLE)).collect();
        return "Stored data to table " + destTableName;
    }

    public Variant ingest_data(Session session, String resourceId) {
        var table = new DefaultKeyValueTable(session, "STATE.APP_CONFIGURATION");
        Map<String, Variant> configuration = table.fetch("config").asMap();
        String destDatabaseName = configuration.get("destination_database").toString();

        String destTableName = destDatabaseName + "." + "DEST_SCHEMA" + "." + resourceId.replaceAll("/","_");

        var split = resourceId.split("/");
        String organization = split[0];
        String repository = split[1];

        String url = String.format("https://api.github.com/repos/%s/%s/issues", organization, repository);

        while (true) {
            FetchResult result = fetch_single_batch(session, url, destTableName);
            session.sql(String.format("GRANT SELECT ON TABLE %s TO APPLICATION ROLE %s", destTableName, APP_ROLE)).collect();
            if (result.getNextPageUrl() == null) {
                break;
            }
            url = result.getNextPageUrl();
        }

        return new Variant("Stored data to table " + destTableName);
    }

    class FetchResult {
        private final String nextPageUrl;

        FetchResult(String nextPageUrl) {
            this.nextPageUrl = nextPageUrl;
        }

        public String getNextPageUrl() {
            return nextPageUrl;
        }
    }

    public FetchResult fetch_single_batch(Session session, String url, String destTableName) {

        HttpResponse<String> response = new GithubApiHttpClient().get(url);

        String body = response.body();
        GithubIssue[] issues;
        try {
            issues = objectMapper.readValue(body, GithubIssue[].class);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Cannot parse json", e);
        }

        StructType schema =
            StructType.create(
                new StructField("id", DataTypes.IntegerType),
                new StructField("number", DataTypes.IntegerType),
                new StructField("title", DataTypes.StringType),
                new StructField("state", DataTypes.StringType),
                new StructField("node_id", DataTypes.StringType),
                new StructField("url", DataTypes.StringType)
            );

        List<Row> list = Arrays.stream(issues).map((i) -> Row.create(i.getId(), i.getNumber(), i.getTitle(), i.getState(), i.getNode_id(), i.getUrl())).collect(Collectors.toList());
        session.createDataFrame( list.toArray(new Row[]{}), schema).write().mode(SaveMode.Append).saveAsTable(destTableName);;
        return parseLinks(response);
    }

    private FetchResult parseLinks(HttpResponse<String> response) {
        return response.headers()
                .firstValue("Link")
                .map((it) -> new LinkHeaderParser().parseLink(it))
                .map((it) -> new FetchResult(it.get("next")))
                .orElse(new FetchResult(null));
    }
}

File: examples-basic/example-github-python-connector/src/snowflake_github_connector/ingestion.py
import re
import requests
from snowflake.snowpark import Session
from snowflake.snowpark.functions import column
from snowflake.snowpark.types import (
    StructType,
    StructField,
    VariantType,
    IntegerType,
    StringType,
)
from snowflake.snowpark.table import WhenMatchedClause, WhenNotMatchedClause

from snowflake_github_connector.common import sink_table_name, APP_ROLE


def get_next_page_link(response: requests.Response):
    page_links = response.headers.get("Link", "")
    for link in page_links.split(","):
        match = re.match('.*<(.+)>; rel="next"', link)
        if match:
            return match.group(1)
    return None


def fetch_single_batch(
    session: Session,
    destination_db: str,
    resource_name: str,
    github_token: str,
    page_link: str = None,
):
    table_name = sink_table_name(destination_db, resource_name)
    url = page_link or f"https://api.github.com/repos/{resource_name}/issues"
    response = requests.get(url, headers={"Authorization": f"Bearer {github_token}"})
    response.raise_for_status()

    # Save data
    source = session.create_dataframe(
        [{"raw": r} for r in response.json()],
        schema=StructType([StructField("raw", VariantType())]),
    )
    target = session.table(table_name)
    target.merge(
        source,
        target["raw"]["id"] == source["raw"]["id"],
        [
            WhenMatchedClause().update({"raw": source["raw"]}),
            WhenNotMatchedClause().insert({"raw": source["raw"]}),
        ],
    )

    # Return state for next iteration
    return {
        "table_name": table_name,
        "rows_count": session.table(table_name).count(),
        "extras": {
            "next_page_link": get_next_page_link(response),
            "current_page_link": url,
        },
    }


def create_sink_table_and_view(destination_db: str, resource_id: str, session: Session):
    # Create sink table
    table_name = sink_table_name(destination_db, resource_id)
    view_name = table_name + "_FLATTENED"
    session.create_dataframe([], schema=StructType([StructField("raw", VariantType())])).write.mode(
        "ignore"
    ).save_as_table(table_name)
    # Create flatten data view
    columns = {
        "id": IntegerType(),
        "node_id": StringType(),
        "url": StringType(),
        "repository_url": StringType(),
        "labels_url": StringType(),
        "comments_url": StringType(),
        "events_url": StringType(),
        "html_url": StringType(),
        "number": IntegerType(),
        "state": StringType(),
        "title": StringType(),
        "body": StringType(),
        "user": VariantType(),
    }
    session.table(table_name).select(
        *[column("raw")[k].cast(t).alias(k) for k, t in columns.items()]
    ).create_or_replace_view(view_name)
    # Privileges
    session.sql(f"GRANT SELECT ON TABLE {table_name} TO APPLICATION ROLE {APP_ROLE}").collect()
    session.sql(f"GRANT SELECT ON VIEW {view_name} TO APPLICATION ROLE {APP_ROLE}").collect()
